art/attacks/evasion/patchfool.py

```diff
@@ -0,0 +1,258 @@
+                "The framework-specific implementation only supports framework-specific preprocessing."
+            )
+
+        super().__init__(estimator=estimator)
+
+        self.alpha = alpha
+        self.max_iter = max_iter
+        self.batch_size = batch_size
+        self.learning_rate = learning_rate
+        self.step_size = step_size
+        self.step_size_decay = step_size_decay
+        self.patch_layer = patch_layer
+        self.random_start = random_start
+        self.skip_att_loss = skip_att_loss
+        self._check_params()
+
+    def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> np.ndarray:
+        """
+        Generate adversarial samples and return them in an array.
+
+        :param x: Source samples.
+        :param y: Target labels.
+        :return: Adversarial examples.
+        """
+        import torch
+        from torchvision.transforms import functional as F
+
+        nb_samples = x.shape[0]
+        x_adv = [None] * nb_samples
+        nb_batches = int(np.ceil(nb_samples / float(self.batch_size)))
+
+        if y is None:
+            y = get_labels_np_array(self.estimator.predict(x, batch_size=self.batch_size))
+
+        for idx in range(nb_batches):
+
+            begin, end = idx * self.batch_size, min((idx + 1) * self.batch_size, nb_samples)
+
+            x_batch = torch.from_numpy(x[begin:end]).to(dtype=torch.float32)
+            y_batch = torch.from_numpy(y[begin:end]).to(dtype=torch.float32)
+
+            x_adv[begin:end] = self._generate_batch(x_batch, y_batch).cpu().detach().numpy()
+
+        return np.array(x_adv)
+
+    def _generate_batch(self, x: "torch.Tensor", y: Optional["torch.Tensor"] = None) -> "torch.Tensor":
+        """
+        TODO
+        """
+        import torch
+        from torch.nn import functional as F
+
+        x = x.to(self.estimator.device)
+        y = y.to(self.estimator.device)
+
+        patch_list = self._get_patch_index(x, layer=self.patch_layer)
+
+        patch_size = self.estimator.patch_size
+        mask = torch.zeros(x.shape).to(self.estimator.device)
+
+        for n, patch_idx in enumerate(patch_list):
+            row = (patch_idx // (x.shape[2] // patch_size)) * patch_size
+            col = (patch_idx % (x.shape[2] // patch_size)) * patch_size
+            mask[n, :, row : row + patch_size, col : col + patch_size] = 1
+
+        x_adv = torch.clone(x).to(self.estimator.device)
+        x_adv = torch.mul(x_adv, 1 - mask)
+
+        if self.random_start:
+            perturbation = torch.rand(x.shape).to(self.estimator.device)
+        else:
+            perturbation = torch.zeros(x.shape).to(self.estimator.device)
+        perturbation.requires_grad = True
+
+        optim = torch.optim.Adam([perturbation], lr=self.learning_rate)
+        scheduler = torch.optim.lr_scheduler.StepLR(optim, step_size=self.step_size, gamma=self.step_size_decay)
+
+        for i_max_iter in tqdm(range(self.max_iter)):
```

Comment:
The variable `i_max_iter ` seems not be used, you can replace it with `_` to avoid the CodeQL alert.